File: Q00101-Q00200/00167-Two-Sum-II-Input-array-is-sorted/java00167/Solution2.java
package java00167;

import java.util.HashMap;
import java.util.Map;

/**
 * @Author: alton
 * @Date: Created in 5/30/21 9:02 AM
 * @Description:
 * 167. Two Sum II - Input array is sorted #57
 *
 * Given an array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number.
 *
 * Return the indices of the two numbers (1-indexed) as an integer array answer of size 2, where 1 <= answer[0] < answer[1] <= numbers.length.
 *
 * The tests are generated such that there is exactly one solution. You may not use the same element twice.
 *
 *
 *
 * Example 1:
 *
 * Input: numbers = [2,7,11,15], target = 9
 * Output: [1,2]
 * Explanation: The sum of 2 and 7 is 9. Therefore index1 = 1, index2 = 2.
 * Example 2:
 *
 * Input: numbers = [2,3,4], target = 6
 * Output: [1,3]
 * Example 3:
 *
 * Input: numbers = [-1,0], target = -1
 * Output: [1,2]
 *
 *
 * Constraints:
 *
 * 2 <= numbers.length <= 3 * 104
 * -1000 <= numbers[i] <= 1000
 * numbers is sorted in non-decreasing order.
 * -1000 <= target <= 1000
 * The tests are generated such that there is exactly one solution.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * Runtime: 2 ms, faster than 25.11% of Java online submissions for Two Sum II - Input array is sorted.
 * Memory Usage: 39 MB, less than 64.40% of Java online submissions for Two Sum II - Input array is sorted.
 */
class Solution2 {
    public int[] twoSum(int[] numbers, int target) {
        Map<Integer, Integer> help = new HashMap<>();

        for (int i = 1; i < numbers.length + 1; i++) {
            if (help.containsKey(numbers[i - 1])) {
                int[] res = new int[2];
                res[0] = help.get(numbers[i - 1]);
                res[1] = i;
                return res;
            } else {
                help.put(target - numbers[i - 1], i);
            }
        }

        return new int[]{-1, -1};
    }
}

File: Q01401-Q01500/01449-Form-Largest-Integer-With-Digits-That-Add-up-to-Target/cpp01449/Solution2.cpp
// Runtime: 24 ms, faster than 72.58% of C++ online submissions for Form Largest Integer With Digits That Add up to Target.
// Memory Usage: 19.4 MB, less than 72.18% of C++ online submissions for Form Largest Integer With Digits That Add up to Target.

class Solution {
public:
    string largestNumber(vector<int>& cost, int target) {

        vector<vector<int>> help(10, vector<int>(target + 1, INT_MIN));
        vector<vector<int>> from(10, vector<int>(target + 1));

        help[0][0] = 0;

        for (int i = 0; i < 9; i++) {
            int c = cost[i];

            for (int j = 0; j <= target; j++) {

                if (j < c || help[i][j] > help[i + 1][j -c] + 1) {
                    help[i + 1][j] = help[i][j];
                    from[i + 1][j] = j;
                } else {
                    help[i + 1][j] = help[i + 1][j - c] + 1;
                    from[i + 1][j] = j - c;
                }
            }
        }


        if (help[9][target] < 0) {
            return "0";
        }

        int i = 9, j = target;
        string res;

        while (i > 0) {

            if (j == from[i][j]) {
                i--;
            } else {
                res += '0' + i;
                j = from[i][j];
            }
        }

        return res;
    }
};